x-pack/apm-server/aggregation/spanmetrics/aggregator.go

```diff
@@ -0,0 +1,243 @@
+				"publishing span metrics failed: %s", err,
+			)
+		}
+	}
+	return nil
+}
+
+// Stop stops the Aggregator if it is running, waiting for it to flush any
+// aggregated metrics and return, or for the context to be cancelled.
+//
+// After Stop has been called the aggregator cannot be reused, as the Run
+// method will always return immediately.
+func (a *Aggregator) Stop(ctx context.Context) error {
+	a.stopMu.Lock()
+	select {
+	case <-a.stopped:
+	case <-a.stopping:
+		// Already stopping/stopped.
+	default:
+		close(a.stopping)
+	}
+	a.stopMu.Unlock()
+
+	select {
+	case <-a.stopped:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return nil
+}
+
+func (a *Aggregator) publish(ctx context.Context) error {
+	// We hold a.mu only long enough to swap the spanMetrics. This will
+	// be blocked by spanMetrics updates, which is OK, as we prefer not
+	// to block spanMetrics updaters. After the lock is released nothing
+	// will be accessing a.inactive.
+	a.mu.Lock()
+	a.active, a.inactive = a.inactive, a.active
+	a.mu.Unlock()
+
+	size := a.inactive.size()
+	if size == 0 {
+		a.config.Logger.Debugf("no span metrics to publish")
+		return nil
+	}
+
+	now := time.Now()
+	metricsets := make([]transform.Transformable, 0, size)
+	for hash, metrics := range a.inactive.m {
+		metricset := makeMetricset(now, metrics.key, metrics.count, metrics.sum)
+		metricsets = append(metricsets, &metricset)
+		delete(a.inactive.m, hash)
+	}
+
+	a.config.Logger.Debugf("publishing %d metricsets", len(metricsets))
+	return a.config.Report(ctx, publish.PendingReq{
+		Transformables: metricsets,
+		Trace:          true,
+	})
+}
+
+func (a *Aggregator) AggregateTransformables(in []transform.Transformable) {
+	for _, tf := range in {
+		if span, ok := tf.(*model.Span); ok &&
+			span.DestinationService != nil &&
+			span.DestinationService.Resource != nil {
+			key := aggregationKey{
+				serviceEnvironment: span.Metadata.Service.Environment,
+				serviceName:        span.Metadata.Service.Name,
+				resource:           *span.DestinationService.Resource,
+			}
+			duration := time.Duration(span.Duration * float64(time.Millisecond))
+			metrics := spanMetrics{
+				count: span.RepresentativeCount,
+				sum:   duration.Milliseconds(),
+				key:   key,
+			}
+			a.active.storeOrUpdate(metrics)
+		}
+	}
+}
+
+type metricsBuffer struct {
+	// TODO might need a size cap
+	m  map[uint64]spanMetrics
+	mu sync.RWMutex
+}
+
+func newMetricsBuffer() *metricsBuffer {
+	return &metricsBuffer{
+		m: make(map[uint64]spanMetrics),
+	}
+}
+
+func (mb *metricsBuffer) load(key uint64) (spanMetrics, bool) {
+	mb.mu.RLock()
+	defer mb.mu.RUnlock()
+	result, ok := mb.m[key]
+	return result, ok
+}
+
+func (mb *metricsBuffer) size() int {
+	// TODO might not need to lock
+	mb.mu.RLock()
+	defer mb.mu.RUnlock()
+	size := len(mb.m)
+	return size
+}
+
+func (mb *metricsBuffer) storeOrUpdate(value spanMetrics) {
+	hash := value.key.hash()
+	if data, ok := mb.load(hash); ok {
```

Comment:
So it is about correctness of simplicity? I went with the classical implementation from the books: read lock to read, then write lock to write. Did I miss an edge case for which it doesn't work? How is that not simple? 
Would it help if the read is not in a separated method?

Reply:
There's a correctness issue mentioned above about using only the hash for the map key.

The comment directly above yours is mostly about simplicity; I assumed you didn't want to copy the txmetrics code because it was complicated (which would be fair). In addition to that, the current implementation using read and write locks is not correct: see https://github.com/elastic/apm-server/pull/4077#discussion_r472060292.

> Would it help if the read is not in a separated method?

No. Bottom line: you can't fetch a value with a read lock, release the read lock, and then take a write lock and expect the value to be the same.

Aside from that, as mentioned in https://github.com/elastic/apm-server/pull/4077#discussion_r471935416 you end up taking the write lock _anyway_, so you may as well just create a single lock around everything. Performance wise, that's what it amounts to anyway.

To move things along, how about changing to a simple `map[aggregationKey]spanMetrics` protected with a `sync.Mutex` for both reads and writes, and we can follow up with a more optimised approach.